File: leetcode/0033-search-in-rotated-sorted-array/Solution.java
class Solution {
    public int search(int[] nums, int target) {
        return search(nums, 0, nums.length-1, target);
    }
    
    private int search(int[] nums, int lo, int hi, int target) {
        if (lo > hi)    // not found
            return -1;
        
        int mid = lo + (hi - lo) / 2;
        
        if (nums[mid] == target)    // found
            return mid;
        
        if (nums[lo] <= nums[mid]) {    // nums[lo...mid] is sorted
            if (nums[lo] <= target && target <= nums[mid])   // target lies in nums[lo...mid]
                return search(nums, lo, mid-1, target);
            
            return search(nums, mid+1, hi, target);
        }
        
        
        // nums[mid...hi] is sorted
        if (nums[mid] <= target && target <= nums[hi])
            return search(nums, mid+1, hi, target);

        return search(nums, lo, mid-1, target);
    }
}

File: leetcode/0033-search-in-rotated-sorted-array/solution.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        return self.searchUtil(nums, 0, len(nums) - 1, target)
    
    def searchUtil(self, nums: List[int], lo: int, hi: int, target: int) -> int:
        if lo > hi:
            return -1
        
        mid = lo + (hi - lo) // 2
        
        if nums[mid] == target:
            return mid
        
        if nums[mid] < nums[hi]:
            if nums[mid] < target and target <= nums[hi]:
                return self.searchUtil(nums, mid + 1, hi, target)
            else:
                return self.searchUtil(nums, lo, mid - 1, target)
        else:
            if nums[lo] <= target and target < nums[mid]:
                return self.searchUtil(nums, lo, mid - 1, target)
            else:
                return self.searchUtil(nums, mid + 1, hi, target)